pkg/kubelet/dockershim/network/cni/cni.go

```diff
@@ -235,19 +242,23 @@ func (plugin *cniNetworkPlugin) setDefaultNetwork(n *cniNetwork) {
 }
 
 func (plugin *cniNetworkPlugin) checkInitialized() error {
-	if plugin.getDefaultNetwork() == nil {
-		return errors.New("cni config uninitialized")
+	defaultCNINetwork := plugin.getDefaultNetwork()
+
+	if defaultCNINetwork == nil {
+		return fmt.Errorf("cni config uninitialized")
 	}
 
-	// If the CNI configuration has the ipRanges capability, we need a PodCIDR assigned
-	for _, p := range plugin.getDefaultNetwork().NetworkConfig.Plugins {
-		if p.Network.Capabilities["ipRanges"] {
-			if plugin.podCidr == "" {
-				return errors.New("no PodCIDR set")
-			}
-			break
-		}
+	netConf, cniNet := defaultCNINetwork.NetworkConfig, defaultCNINetwork.CNIConfig
+
+	caps, err := cniNet.ValidateNetworkList(context.TODO(), netConf)
+	if err != nil {
+		return fmt.Errorf("cni config validation fails : %v", err)
+	}
+
+	if utilslice.ContainsString(caps, ipRangesCapability, nil) && plugin.podCidr == "" {
```

Comment:
@mars1024 I don't think this check is quite right. It's certainly possible that the CNI config would have the capability but that the user has some other IPAM enabled that's not the Kube CIDR allocator. It's also possible that the user forgot to disable the pod CIDR allocator and their CNI plugin handled the pod CIDR some other way (by talking to the apiserver directly).

I think it's possible to argue that if the cluster admin has enabled pod CIDR allocation that they intend for that to be used by the network plugin and thus something like this is valid:

`if plugin.podCidr != "" && !utilslice.ContainsString(caps, ipRangesCapability, nil) {`

but this would also cause a cluster upgrade to fail until the admin updated the CNI configuration file, which might not be OK.

Reply:
Thanks for reviewing ~

I just followed and refactored the original code here, but I think most of the users may not enable ipRanges capability if they have their own way to get pod CIDR, although host-local allow [this](https://github.com/containernetworking/plugins/blob/7ba2bcfeab482eaaa7d6e7e4ec9fbd493c54d4fa/plugins/ipam/host-local/backend/allocator/config.go#L114). IMO, I think runtime config should override fixed config.

Otherwise, I think cni config without ipRanges capability can work normally when pod CIDR is set, we just do not use it... so I don't think we need a check like `if plugin.podCidr != "" && !utilslice.ContainsString(caps, ipRangesCapability, nil)`, it maybe a break change just like you mentioned above.

Do you suppose that we just drop this check to let host-local return error, or just keep the original way? @dcbw

Reply:
@mars1024 sorry, didn't catch that it was just a refactor. I think the original code in https://github.com/kubernetes/kubernetes/pull/64445 was wrong there, but @squeed can argue differently. Casey?

Reply:
@mars1024 after talking to Casey/@squeed I don't think we want to call ValidateNetworkList() on each setup/teardown. This is because it will check that every network plugin exists on-disk *and* execute them to read their version info for validation. That could be a significant set of operations done for every SetUp/TearDown.

Instead, could we put that into getDefaultCNINetwork() when reading the default netowrk config?

Reply:
@dcbw Yes, moving validation to `getDefaultCNINetwork()` seems a better and more *economic* way, but I still have a question, should we move capability check (like ipRanges) to `getDefaultCNINetwork()` either, or just leave it in `checkInitialized` ? cc @squeed

Reply:
The "ipRange capability is added but podCIDR isn't set" check should be in `checkInitialized`, because the podCIDR is assigned asynchronously.

I think that's the only capability check, right?

Reply:
> The "ipRange capability is added but podCIDR isn't set" check should be in `checkInitialized`, because the podCIDR is assigned asynchronously.
> 
> I think that's the only capability check, right?

Yes, I think so, too, the other capabilities are no essential.